Koans/AboutReflection.cs

```diff
@@ -0,0 +1,122 @@
+		// Using reflection we can inspect its types
+		// Hint: we're using a type that we already know about to get the Xunit assembly and all of its types
+
+		Assembly a = typeof(Assert).Module.Assembly;
+
+		Type[] types = a.GetTypes();
+
+		Assert.Contains(typeof(FillMeIn), types);
+	}
+
+
+	[Step(4)]
+	public void ExposingNonPublicConstructor()
+	{
+		// Reflection can be dangerous!
+		// It is possible to access non-public members, types and methods in an assembly
+		// Never execute these methods or set private members as it will likely cause undefined behaviour. 
+		// They are inaccessible for a reason!
+
+		// When inspecting the Assert class (via F12/Go to Definition), we can see that there is a protected constructor without a body
+		// We can use reflection to retrieve non public constructors and execute them
+		// We specify BindingFlags to only retrieve non public, instance constructors.
+		ConstructorInfo[] constructors = typeof(Assert).GetConstructors(BindingFlags.NonPublic | BindingFlags.Instance);
+				
+		// We don't know for sure how many constructors there are in the Assert class when using reflection so we can be safe and check for null before invoking the first element in the array
+		ConstructorInfo constructor = constructors[0];
+
+		// the Invoke method returns object, but because we have retrieved this from reflecting the Assert type, we can infer that its return type is Assert
+        object assertInstance = constructor?.Invoke(null);
+
+		Assert.Equal(typeof(FillMeIn), assertInstance.GetType());
+	}
+
+	[Step(5)]
+	public void ExposingNonPublicFields()
+	{
+		// At the end of this file, there is a class called ACustomType, which contains a private string field that is set by the constructor
+		// We can use reflection to access this instance variable when we otherwise wouldn't be able to
+		
+		string aLocalPrivateString = "this string will be private in ACustomType";
+		string aLocalPublicString = "this string will be public in ACustomType";
+		ACustomType aCustomType = new ACustomType(aLocalPrivateString, aLocalPublicString);
+
+		// Get the Type of our class
+		Type customType = typeof(ACustomType);
+
+		// Get our non-public instance members
+		FieldInfo[] fields = customType.GetFields(BindingFlags.NonPublic | BindingFlags.Instance);
+
+		// Retrieving a public field as we are allowed to
+		Assert.Equal(aLocalPublicString, aCustomType.publicString);
+
+		// We know that there is only one non-public instance field in ACustomType,
+		// so we know the value is what we set as the privateString parameter when we called the constructor above
+		FieldInfo privateField = fields[0];
+
+		// as in the previous step, our reflected value is of type object, which we know from the class below is a string
+		object privateFieldValue = privateField.GetValue(aCustomType);
+
+		Assert.Equal(typeof(FillMeIn), privateFieldValue.GetType());
+		Assert.Equal(FILL_ME_IN, privateFieldValue);
+	}	
```

Comment:
Should we add one example where reflection is really useful?
Do you have an interesting idea?
I thought about accessing custom attributes.

Reply:
In my mind the most common/obvious use for reflection would be paired with generics to make a serialiser (CSV, JSON etc). I'd ideally want to add the useful reflection example in the generics topic. Thoughts?

There's a lot of potential for encouraging pretty dangerous anti-patterns, so I tried to keep it fairly high level. Eg. I intentionally omitted any mention of loading external assemblies as I wouldn't want anyone new to C# writing in RCE vulnerabilities 😄